tm2/pkg/sdk/helpers.go

```diff
@@ -10,17 +10,34 @@ import (
 var isAlphaNumeric = regexp.MustCompile(`^[a-zA-Z0-9]+$`).MatchString
 
 func (app *BaseApp) Check(tx Tx) (result Result) {
-	return app.runTx(RunTxModeCheck, nil, tx)
+	ctx := app.getContextForTx(RunTxModeCheck, nil)
+
+	return app.runTx(ctx, tx)
 }
 
 func (app *BaseApp) Simulate(txBytes []byte, tx Tx) (result Result) {
-	return app.runTx(RunTxModeSimulate, txBytes, tx)
+	ctx := app.getContextForTx(RunTxModeSimulate, txBytes)
+
+	return app.runTx(ctx, tx)
 }
 
-func (app *BaseApp) Deliver(tx Tx) (result Result) {
-	return app.runTx(RunTxModeDeliver, nil, tx)
+func (app *BaseApp) Deliver(tx Tx, ctxFns ...ContextFn) (result Result) {
+	ctx := app.getContextForTx(RunTxModeDeliver, nil)
+
+	for _, ctxFn := range ctxFns {
+		if ctxFn == nil {
+			continue
+		}
+
+		ctx = ctxFn(ctx)
+	}
+
+	return app.runTx(ctx, tx)
 }
 
+// ContextFn is the custom execution context builder
```

Comment:
This feature increases complexity, so please add a comment explaining why it's needed or provide example use cases.

Reply:
```suggestion
// ContextFn is the custom execution context builder.
// It can be used to add custom metadata when replaying transactions
// during InitChainer or in the context of a unit test.
```